### PR TITLE
[BREAKING DevTools] Adds tools for testing async trees, such as Relay ones.

### PR DESCRIPTION
**Pull request by alloy**
* Consolidate existing dev tools into single dir. Including the previously exported `ClientRouter`, because that uses our schema and would implicitly pull that into our JS bundles, so we should revisit that when we really need it somewhere at runtime.
* Add [`renderUntil`](https://github.com/artsy/reaction/pull/1345/files#diff-99ca552957c29b34c115ada9663e43b7) which is used to test an async tree.
* Add [`renderRelayTree`](
https://github.com/artsy/reaction/pull/1345/files#diff-672b34f9f1a8662fbbc556b68ef79a71) which is used to test Relay trees.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -37,7 +37,7 @@
     "stylelint-project": " stylelint -- 'src/**/*.{ts,tsx}'",
     "prettier-project": "yarn run prettier-write 'src/**/*.{ts,tsx}'",
     "prettier-write": "yarn run prettier --write",
-    "relay": "relay-compiler --src ./src --schema data/schema.graphql --language typescript --artifactDirectory ./src/__generated__",
+    "relay": "relay-compiler --src ./src --schema data/schema.graphql --language typescript --artifactDirectory ./src/__generated__ --exclude '**/node_modules/****/__mocks__/**,**/__generated__/**'",
```
Comment (damassi) [inline package.json]: What does `****` and `**, **` do?

File: src/Apps/Order/Routes/__tests__/Status.test.tsx [hunk 2]
```diff
@@ -1,32 +1,56 @@
...
-import { HeadProvider } from "react-head"
-import { Status_order } from "../../../../__generated__/Status_order.graphql"
+import { graphql } from "react-relay"
 import {
+  mockResolver,
   OrderWithShippingDetails,
   PickupOrder,
 } from "../../../../Apps/__test__/Fixtures/Order"
-import { StatusRoute } from "../Status"
+import { MockBoot, renderRelayTree, UntilCallback } from "../../../../DevTools"
```
Comment (damassi) [inline src/Apps/Order/Routes/__tests__/Status.test.tsx]: Once TypeScript support is fixed in our tests these auto imports will be absolute

**Comment by artsyit** [comment]
Deploy preview for *artsy-reaction* ready!

Built with commit af7b25ef5fc79c1646542073d4c08a3b66153fbc

https://deploy-preview-1345--artsy-reaction.netlify.com

File: package.json (hunk 1 above)
Comment (alloy) [reply package.json]: Hmmmm, this is copy-🍝 from the `relay-compiler` source which was excluding tests, but I think I missed inserting a comma between `****`, these should be 3 exclude patterns.

Great catch! 👌

File: src/Apps/Order/Routes/__tests__/Status.test.tsx [hunk 3]
```diff
@@ -1,32 +1,56 @@
...
+        }
+      `,
+      mockResolvers: mockResolver(order),
+      wrapper: renderer => (
+        <MockBoot breakpoint="xs" headTags={headTags}>
+          {renderer}
+        </MockBoot>
+      ),
+    })
+  }
```
Comment (damassi) [inline src/Apps/Order/Routes/__tests__/Status.test.tsx]: 👍

File: src/Apps/Order/Routes/__tests__/Status.test.tsx [hunk 4]
```diff
@@ -1,32 +1,56 @@
...
-  const getWrapper = (order: Status_order, headTags = []) =>
-    mount(
-      <HeadProvider headTags={headTags}>
-        <StatusRoute order={order} />
-      </HeadProvider>
-    )
+  const getWrapper = (
+    order: any,
+    headTags: Array<React.ReactElement<any>> = [],
+    until: UntilCallback = null
```
Comment (damassi) [inline src/Apps/Order/Routes/__tests__/Status.test.tsx]: What does `UntilCallback` do?

File: src/DevTools/MockRelayRenderer.tsx [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+      Component={Component}
+      mockResolvers={mockResolvers}
+      query={query}
+    />
+  )
+  return renderUntil<P, S, C>(
+    until || (tree => !tree.find(`.${LoadingClassName}`).length),
+    wrapper ? wrapper(renderer) : renderer
+  )
+}
```
Comment (alloy) [inline src/DevTools/MockRelayRenderer.tsx]: I should move this function out into its own module.

File: src/setup_jest.ts [hunk 6]
```diff
@@ -9,6 +9,12 @@ track.mockImplementation(y => x => x)
 
 jest.mock("react-sizeme", () => jest.fn(c => d => d))
 
+/**
+ * We want each test to have assertions, otherwise it’s too easy to write async
+ * tests that never end up making any, leading to false positives.
+ */
+beforeEach(() => expect.hasAssertions())
```
Comment (alloy) [inline src/setup_jest.ts]: We should do this in all our test suites. /cc @orta

File: src/Styleguide/Components/__tests__/ArtistBio.test.tsx [hunk 7]
```diff
@@ -19,7 +19,8 @@ describe("ArtistBio", () => {
     expect(wrapper.find("ReadMore").length).toBe(1)
   })
 
-  it("triggers callback when clicked", done => {
+  // TODO: Chris, this test needs finishing.
+  xit("triggers callback when clicked", done => {
```
Comment (alloy) [inline src/Styleguide/Components/__tests__/ArtistBio.test.tsx]: cc @damassi

File: src/Styleguide/Components/__tests__/CountrySelect.test.tsx [hunk 8]
```diff
@@ -3,7 +3,8 @@ import React from "react"
 import { CountrySelect } from "../CountrySelect"
 
 describe("CountrySelect", () => {
-  it("triggers callback on change", done => {
+  // TODO: Chris, this test needs finishing.
+  xit("triggers callback on change", done => {
```
Comment (alloy) [inline src/Styleguide/Components/__tests__/CountrySelect.test.tsx]: cc @damassi

File: src/Apps/__stories__/Apps.story.tsx [hunk 9]
```diff
@@ -8,15 +8,15 @@ import { routes as collectRoutes } from "../Collect/routes"
 storiesOf("Apps", module)
   .add("Artwork Page", () => {
     return (
-      <ClientRouter
+      <MockRouter
```
Comment (damassi) [inline src/Apps/__stories__/Apps.story.tsx]: 👍 I guess if we need a quick client-side router its easy enough to use `buildClientRouter` directly

File: tslint.json [hunk 10]
```diff
@@ -21,7 +21,8 @@
       { "artifactDirectory": "__generated__", "makeRelative": false }
     ],
     "trailing-comma": false,
-    "variable-name": [true, "ban-keywords", "allow-leading-underscore"]
+    "variable-name": [true, "ban-keywords", "allow-leading-underscore"],
+    "jsdoc-format": false
```
Comment (alloy) [inline tslint.json]: I’m excluding the asterisks in `@example` docs because of https://github.com/Microsoft/TypeScript/issues/23517

File: src/Apps/Artwork/Utils/__tests__/filterLocations.test.tsx [hunk 11]
```diff
@@ -3,12 +3,12 @@ import { filterLocations } from "../filterLocations"
...
   it("returns null if no array of locations is given", () => {
     const filtered = filterLocations(null)
-    expect(filtered).toBeNull
+    expect(filtered).toBeNull()
   })
 
   it("returns null if empty array of locations is given", () => {
     const filtered = filterLocations([])
-    expect(filtered).toBeNull
+    expect(filtered).toBeNull()
```
Comment (alloy) [inline src/Apps/Artwork/Utils/__tests__/filterLocations.test.tsx]: Such Ruby! 😅

File: src/Apps/Order/Routes/__tests__/Status.test.tsx (hunk 4 above)
Comment (alloy) [reply src/Apps/Order/Routes/__tests__/Status.test.tsx]: See the doc in `renderUntil` https://github.com/artsy/reaction/pull/1345/files#diff-99ca552957c29b34c115ada9663e43b7R33

File: src/Artsy/Relay/renderWithLoadProgress.tsx [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+      //   // TODO: Even though this code path is reached, the retry button keeps spinning. iirc it _should_ disappear when
+      //   //      `onRetry` on the instance is unset.
+      //   //
+      //   // This will re-use the native view first created in the renderFailure callback, which means it can
+      //   // continue its ‘retry’ animation.
+      //   return <LoadFailureView style={{ flex: 1 }} />
+      // } else {
+      //   retrying = true
+      //   return <LoadFailureView onRetry={retry} style={{ flex: 1 }} />
+      // }
```
Comment (alloy) [inline src/Artsy/Relay/renderWithLoadProgress.tsx]: @damassi cc

File: src/DevTools/__test__/MockRelayRenderer.test.tsx [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+      expect(tree.html()).toEqual(
+        renderToString(
+          <div>
+            <img src="http://test/image.jpg" />
+            <div>Mona Lisa</div>
+          </div>
+        )
+      )
+      done()
+    }, 10)
```
Comment (damassi) [inline src/DevTools/__test__/MockRelayRenderer.test.tsx]: Will this ever unexpectedly fail in the future?

File: src/Styleguide/Components/RecentlyViewed.tsx [hunk 14]
```diff
@@ -105,13 +106,7 @@ export const RecentlyViewedQueryRenderer = () => {
...
               }
             `}
-            render={({ props }) => {
-              if (props) {
-                return <RecentlyViewedFragmentContainer me={props.me} />
-              } else {
-                return null
-              }
-            }}
+            render={renderWithLoadProgress(RecentlyViewedFragmentContainer)}
```
Comment (damassi) [inline src/Styleguide/Components/RecentlyViewed.tsx]: Noticed a few spots lately that could use this

File: src/Apps/Order/Routes/__tests__/Payment.test.tsx [hunk 15]
```diff
@@ -226,8 +226,6 @@ describe("Payment", () => {
     fillAddressForm(paymentRoute, validAddress)
 
     paymentRoute.find(ContinueButton).simulate("click")
-
-    expect.hasAssertions()
```
Comment (zephraph) [inline src/Apps/Order/Routes/__tests__/Payment.test.tsx]: Is this just some unrelated cleanup?

File: src/DevTools/MockRelayRenderer.tsx [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+   ```
+ *
+ */
+export function renderRelayTree<
+  P = {},
+  S = {},
+  C extends React.Component = React.Component
+>(
+  params: MockRelayRendererProps & {
+    until?: UntilCallback<P, S, C>
```
Comment (damassi) [inline src/DevTools/MockRelayRenderer.tsx]: `UntilCallback` usage makes total sense but I'm wondering if there might be a slightly more descriptive name for it? Since it **a)** involves a common situation around nested `QueryRenderer`'s and **b)** could be used in other ways.

File: src/Apps/__stories__/Apps.story.tsx (hunk 9 above)
Comment (alloy) [reply src/Apps/__stories__/Apps.story.tsx]: Probably yeah 👍 and likely best to do that once before we further abstract it

File: src/DevTools/__test__/MockRelayRenderer.test.tsx [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+
+const query = graphql`
+  query MockRelayRendererQuery {
+    artwork(id: "mona-lisa") {
+      ...MockRelayRenderer_artwork
+    }
+  }
+`
+
+function renderToString(element: React.ReactElement<any>) {
```
Comment (damassi) [inline src/DevTools/__test__/MockRelayRenderer.test.tsx]: Thoughts on `React.ReactElement<any>` vs `JSX.Element` in common code? The former opens the doors to more safety later on and so seems better to use but was wondering your thoughts.

File: src/DevTools/__test__/MockRelayRenderer.test.tsx (hunk 13 above)
Comment (alloy) [reply src/DevTools/__test__/MockRelayRenderer.test.tsx]: Timers aren’t mocked by default: https://jestjs.io/docs/en/timer-mocks.html

What part of he code are you thinking of? I don’t really see a reason for this to fail (other than the code being broken).

File: src/Styleguide/Components/RecentlyViewed.tsx (hunk 14 above)
Comment (alloy) [reply src/Styleguide/Components/RecentlyViewed.tsx]: Ah right, I had intended to cc you on me adding this to `RecentlyViewed`. Yeah we should use it in more places 👍

File: src/Apps/Order/Routes/__tests__/Payment.test.tsx (hunk 15 above)
Comment (alloy) [reply src/Apps/Order/Routes/__tests__/Payment.test.tsx]: Yup, see this commit https://github.com/artsy/reaction/pull/1345/commits/1dec03116e2acce1092f644861245dd8f413bf49

I made our setup expect that every test should have an assertion, a few test bugs came out of that.

File: src/DevTools/__test__/MockRelayRenderer.test.tsx (hunk 13 above)
Comment (damassi) [reply src/DevTools/__test__/MockRelayRenderer.test.tsx]: Ah, was thinking about `jest.useFakeTimers()` and the seemingly arbitrary `10ms` use here.

File: src/Artsy/Relay/renderWithLoadProgress.tsx (hunk 12 above)
Comment (zephraph) [reply src/Artsy/Relay/renderWithLoadProgress.tsx]: Are we leaving this for follow up?

File: src/Apps/Order/Routes/__tests__/Status.test.tsx [hunk 18]
```diff
@@ -1,32 +1,56 @@
...
   it("should should have a title containing status", async () => {
-    const headTags = []
-    getWrapper(OrderWithShippingDetails, headTags)
-    // @ts-ignore
-    expect(render(headTags).text()).toBe("Order status | Artsy")
+    const headTags: Array<React.ReactElement<any>> = []
+    await getWrapper(
+      OrderWithShippingDetails,
+      headTags,
+      () => headTags.length > 0
```
Comment (alloy) [inline src/Apps/Order/Routes/__tests__/Status.test.tsx]: @zephraph @damassi I now realise I need to dive into this a bit more. I had previously assumed that the title not showing up immediately was due to how `react-head` works, but am now thinking it’s because of our `Responsive` component. I thought that with an initial breakpoint set the `Responsive` component would immediately render, but maybe not?

File: src/Artsy/Relay/renderWithLoadProgress.tsx (hunk 12 above)
Comment (alloy) [reply src/Artsy/Relay/renderWithLoadProgress.tsx]: Yeah. I probably should have linked to the comment above https://github.com/artsy/reaction/pull/1345/files/af7b25ef5fc79c1646542073d4c08a3b66153fbc#diff-d997111e63b21d4529fb81271cfe5c47R18

File: src/DevTools/MockRelayRenderer.tsx (hunk 16 above)
Comment (alloy) [reply src/DevTools/MockRelayRenderer.tsx]: Am I correct in understanding that you are suggesting to rename the parameters from `until` to `renderUntil`? And also rename the callback type from `UntilCallback` to `RenderUntilCallback`?

File: src/DevTools/__test__/MockRelayRenderer.test.tsx (hunk 13 above)
Comment (alloy) [reply src/DevTools/__test__/MockRelayRenderer.test.tsx]: Yeah that value is arbitrary, but shouldn’t break unless it is brought down to a very low value.

File: src/DevTools/__test__/MockRelayRenderer.test.tsx (hunk 17 above)
Comment (alloy) [reply src/DevTools/__test__/MockRelayRenderer.test.tsx]: Ah, a `ReactNode` is anything that React knows how to render, including strings, which doesn’t make sense for our case. Ok, going with `JSX.Element` 👍

File: src/DevTools/MockRelayRenderer.tsx (hunk 16 above)
Comment (damassi) [reply src/DevTools/MockRelayRenderer.tsx]: With TS this isn't _so_ much of an issue because one can hover and read description, but not everyone uses TS -- so was thinking of some way of better describe what's happening here. `until` -> `waitUntil` -> `renderUntil`. Haha not too sure... perhaps the variety of its uses warrants a simple name.

File: src/Apps/Order/Routes/__tests__/Status.test.tsx (hunk 18 above)
Comment (damassi) [reply src/Apps/Order/Routes/__tests__/Status.test.tsx]: There's a lot going on inside of react-head considering how it works. What makes you think the issue is with Responsive?

Comment (alloy) [reply src/Apps/Order/Routes/__tests__/Status.test.tsx]: Yeah the new context API is built around Fiber, so while I haven’t traced it I’m pretty sure it’s async https://github.com/facebook/react/blob/master/packages/react-reconciler/src/ReactFiberNewContext.js

Hrmm 🤔 I think we should follow up in another PR, if we find a way to deal with that at all.

Comment (alloy) [reply src/Apps/Order/Routes/__tests__/Status.test.tsx]: I’m gonna leave it at this for now https://twitter.com/alloy/status/1047199392581767168, we’ll see if we hear back or if we can come up with another solution after sleeping on it.

In any case, this shouldn’t block the PR.

File: src/DevTools/MockRelayRenderer.tsx (hunk 16 above)
Comment (alloy) [reply src/DevTools/MockRelayRenderer.tsx]: I didn’t go with `waitUntil` in the first place because this implementation is very tied to rendering React components and so didn’t want to be too ambiguous about that. I’ll rename things as per my above understanding for now.

File: src/Apps/Order/Routes/__tests__/Status.test.tsx (hunk 18 above)
Comment (alloy) [reply src/Apps/Order/Routes/__tests__/Status.test.tsx]: > What makes you think the issue is with Responsive?

I recall not seeing log statements showing up when I expected them to. I should probably just create a test that uses a context to verify.

File: src/DevTools/MockRelayRenderer.tsx (hunk 16 above)
Comment (damassi) [reply src/DevTools/MockRelayRenderer.tsx]: Sounds good -- i think `renderUntil` is a little more clear.

**Comment by alloy** [comment]
@zephraph FYI This is complete from my pov.
